libs/frontend/feature-build/src/lib/create-code-editor-view.ts

```diff
@@ -0,0 +1,67 @@
+import {
+  editorStateTextContent,
+  prosemirrorSetup,
+} from '@pubstudio/frontend/util-edit-text'
+import hljs from 'highlight.js/lib/core'
+import { keymap } from 'prosemirror-keymap'
+import { Plugin, Selection } from 'prosemirror-state'
+import { EditorView } from 'prosemirror-view'
+import { highlightPlugin } from './prosemirror-highlighthjs-plugin'
+
+const codeEditorPlugins: Plugin[] = [
+  keymap({
+    // Insert 2 spaces in the selected position when tab is pressed.
+    Tab: (state, dispatch) => {
+      dispatch?.(state.tr.insertText('  ').scrollIntoView())
+      return true
+    },
+  }),
+  highlightPlugin(hljs),
+]
+
+export function createCodeEditorView(
+  code: string,
+  container: HTMLElement,
+  onChange: (code: string) => void,
+): EditorView | undefined {
+  if (!container) {
+    return undefined
+  }
+
+  const state = prosemirrorSetup({
+    content: `<pre><code>${code}</code></pre>`,
+    plugins: [...codeEditorPlugins],
+  })
+
+  const view = new EditorView(container, {
+    state,
+    dispatchTransaction(transaction) {
+      let newState = view.state.apply(transaction)
+      const { textContent } = newState.doc
+      if (newState.doc.content.firstChild?.type.name !== 'code_block') {
+        // When using hljs with ProseMirror editor, the code must be wrapped inside
+        // `<pre><code>...</code></pre>` for highlighting&styling feature to work. If a user
+        // removes all the text from the editor (i.e. select all text then press backspace),
+        // the leading&ending <pre><code> well also be removed, causing the highlighting&styleing
+        // feature to stop working. We need to manually wrap the content inside <pre><code> in such cases.
+
+        // FIXME
+        // 1. The current solution will clear the history because it creates a brand-new editor state.
+        //    Find a way to replace the content in the editor state while keeping everything.
+        // 2. New line characters are ignored if the text is copied from a source other than ProseMirror
+        //    editor. For example, VSCode.
+        newState = prosemirrorSetup({
+          content: `<pre><code>${textContent}</code></pre>`,
+          plugins: [...codeEditorPlugins],
+        })
+        newState.selection = Selection.atEnd(newState.doc)
```

Comment:
Hi @sampullman 

The "Maximum call stack exceeded" error is resolved by copying source code from [prosemirror-highlightjs](https://github.com/b-kelly/prosemirror-highlightjs) and commenting out the unused parts. But there're still some strange issues I didn't managed to solve 😵.

I left some comments here to describe the problem. What I don't understand is why these 2 problems only occurr when the copied text is from a source other than ProseMirror editor 🤔?
For example, if I try to copy some code from VSCode and paste it to the ProseMirror editor in BehaviorModal, these 2 problems will occurr. But if the code (text) is copied&pasted within the same editor, these problems will not show up!

Reply:
I wasn't actually able to reproduce these two issues, unless I misunderstood something. Here's a recording of copy/paste from VSCode:


https://github.com/samatechtw/pubstudio-builder/assets/550982/7ec4d5d6-d890-4b5b-bc40-b5e1dc0cab8d

Reply:
Hi @sampullman 

I just tried it for a couple more times, and I created another issue #11 to keep track of the bugs!
I'll come back and try to fix it when I'm done with my current work!

Reply:
🙏